### PR TITLE
[WIP] - Cache Reaper

### LINKED ISSUE #627: Services that write to disk should detect disk pressure and clean up
Specifically services that write to SSD locations which are a shared resource on the node (e.g. xlang-go-bg).

For example, if a service detects the location it is writing to is near capacity (e.g. 90% used), then it should perform some cleanup actions.

### PR DESCRIPTION
**Pull request by goddardcm**
Fixes #627.

Adds a `creaper` cmd that can be used to monitor and reduce cache directory size. I've annotated the PR inline.

Not sure what approach we'd like to take for getting this binary running within any of the existing docker containers. Likely we would need to create wrapper start scripts that would invoke both the main binary as well as starting creaper in the background. 

Also worth noting that this implementation isn't testable. I wavered a bit on how much effort to put toward making it testable and decided against it at least for the first pass. If we did want some test coverage for the code, i'd advocate for concentrating those efforts on the `clearSpace` function (given a slice of `cachedFile` and other parameters, ensure that the expected files would be deleted) and not really worrying about testing the `getDirectoryInfo` function)

### CODE REVIEW

File: cmd/creaper/logger/logger.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	log15 "gopkg.in/inconshreveable/log15.v2"
+)
+
+type key int
+
+const (
+	loggerKey key = iota
+)
+
+func WithLogger(ctx context.Context, logCtx ...interface{}) context.Context {
```
Comment (goddardcm) [inline cmd/creaper/logger/logger.go]: This is a pattern I've been wanting to experiment with. Can easily be replaced with calls directly into the logging library, but I think storing a `log15.Logger` with the desired current logging context in a `context.Context` is a pretty elegant pattern.

File: cmd/creaper/creaper/creaper.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	clearSpace(ctx, files, totalSize, 4*maxSize/5)
+
+	logger.Info(ctx, "Reap complete.")
+}
+
+func getDirectoryInfo(ctx context.Context, directory string) ([]cachedFile, uint64) {
+	files := make([]cachedFile, 0)
+	totalSize := uint64(0)
+
+	if err := filepath.Walk(directory, func(path string, info os.FileInfo, err error) error {
```
Comment (goddardcm) [inline cmd/creaper/creaper/creaper.go]: Note: This is not terribly performant, but I suspect the usage patterns of this will involve a few large files rather than many many small files. On my laptop, it takes ~20s on a directory with ~500k files. 

From https://golang.org/pkg/path/filepath/#Walk:
> The files are walked in lexical order, which makes the output deterministic but means that for very large directories Walk can be inefficient.

File: cmd/creaper/main.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	"maxSize",
+	"1g",
+	"max cache size (examples: 1048576k, 1024m, 1g)",
+)
+
+var checkFrequency = flag.Duration(
+	"frequency",
+	time.Minute,
+	"frequency with which the creaper should check disk usage",
+)
```
Comment (slimsag) [inline cmd/creaper/main.go]: nit: for consistency with the rest of our code, we usually don't split `flag.*` or `log15.*` calls onto multiple lines. Not a big deal though (and our fault for this not being in a style-guide somewhere) :)

File: cmd/creaper/creaper/creaper.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	ctx = logger.WithLogger(ctx, "fileCount", len(files), "size", totalSize)
+
+	if totalSize < maxSize {
+		logger.Info(ctx, "Cache size within limit")
+		return
+	}
+
+	logger.Info(ctx, "Cache is overgrown. Reaping.")
+
+	clearSpace(ctx, files, totalSize, 4*maxSize/5)
```
[the remark this answers is not in the document]
Comment (nicksnyder) [reply cmd/creaper/creaper/creaper.go]: We do manually delete the whole dir now when it alerts so there is no harm in that. My thought was that if we are automating this we have a chance to be a bit smarter and less heavy handed, but I don't have a strong preference either way. Perhaps one argument for just deleting the whole directory (in addition to it being simpler to implement), is that it would be more clear on alert graphs when the reaper is kicking in.

What do you think @goddardcm?

Do you have an opinion @keegancsmith ?

> 2\. Since it is 4/5th the `maxSize` and not the `totalSize`, isn't it possible the directory grow far past `maxSize` and we only clean up a small amount of the directory until the next reaping? e.g. if `maxSize` is `500mb` and `totalSize` is `2gb` it would take ~2 reapings to get down to acceptable size

My understanding of the code is that after one reaping the directory size would be less than 400mb, which seems like the intended behavior to me.

> which requires an extra Go dependency

I am curious to know what the difference is between what this library provides vs [FileInfo.ModTime()](https://golang.org/pkg/os/#FileInfo), @goddardcm ?

[the remark this answers is not in the document]
Comment (goddardcm) [reply cmd/creaper/creaper/creaper.go]: > > which requires an extra Go dependency
> 
> I am curious to know what the difference is between what this library provides vs [FileInfo.ModTime()](https://golang.org/pkg/os/#FileInfo), @goddardcm ?

tldr: Library provides access to the last _access_ time of a file. The thinking being a file that hasn't been recently accessed should be prioritized for deletion over one that may have been created before but accessed more recently.

The library is a (very simple) platform-agnostic wrapper around the various times that `stat` provides. In this case, I thought it would make the most sense to delete files from the cache that haven't been accessed in the longest time. Really the library just takes care of casting the return value of [`FileInfo.Sys`](https://golang.org/pkg/os/#FileInfo) to the right underlying type (i.e. [windows](https://golang.org/src/os/types_windows.go#L185) vs [linux](https://golang.org/pkg/syscall/#Stat_t)) and exposing the needed bits via a more developer-friendly interface.

File: cmd/creaper/creaper/creaper.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+				"Error walking cache directory",
+				"path", path,
+				"err", err,
+			)
+
+			// Try to proceed.
+			return filepath.SkipDir
+		}
+
+		if info.Mode().IsRegular() {
```
[the remark this answers is not in the document]
Comment (goddardcm) [reply cmd/creaper/creaper/creaper.go]: That's an interesting idea if that's a problem that we have. I could certainly add some code that could keep track of empty directories and remove them (perhaps even if the cache isn't overflowing)?

File: cmd/creaper/main.go (hunk 3 above)
Comment (goddardcm) [reply cmd/creaper/main.go]: Good to know. Fixed.

File: cmd/creaper/main.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+			*cacheDir,
+			*checkFrequency,
+			maxCacheSizeBytes)
+
+		shutdownWaitGroup.Done()
+	}()
+
+	go func() {
+		// Wait for shutdown signal.
+		_ = <-sigtermChan
```
[the remark this answers is not in the document]
Comment (goddardcm) [reply cmd/creaper/main.go]: > What was the motivation behind trying to do graceful shutdowns like this?

This is just a standard practice that I follow when building new long-running services. I've found that adding this ability later on... really never happens, even if it's merited. Having it baked in from the beginning forces future developers to at least consider if they need to make an effort to clean up after themselves.

> (1) may mean we rely on graceful shutdown here and I just haven't seen it

Nope - you're not missing anything.

> (2) may cause someone in the future editing this service to believe we can rely on graceful shutdown

I'll push back a bit here. If the starting point is that the "container or host machine could be terminated at random" then no one can ever rely on graceful shutdown, and be cognizant of that fact, regardless of what shutdown handling code exists. I just consider it good practice to try to shutdown gracefully when you're able to.

File: cmd/creaper/logger/logger.go (hunk 1 above)
Comment (goddardcm) [reply cmd/creaper/logger/logger.go]: Ok saved replying to this one for last... 😄 

> passing around non-request-scoped data with `context` is kind of a big anti-pattern FWIW

I agree 💯. However, I don't think that this implementation incorporates that anti-pattern (other than, pedantically, the [initial creation](https://github.com/sourcegraph/sourcegraph/pull/2135/files#diff-2dc83a0f6e2c3e3f8e70a66972a8472aR71) of the context that has logging goodies included). 

The way I see it, function calls are basically the same as "requests". And so the pattern here is that data associated with a particular function call gets added on by the developer into the context. That way, subsequent calls to log a statement don't need to include a laundry-list of useful context (some of which they may not even have!). In fact, with this pattern, they'll end up logging relevant data from the entire stack trace (which, as a frequent debugger of issues, I think is a great thing!). But, because of the way the context works, a function that adds its own info to a context won't pollute the original context that was passed in to that function call.

> (Sidenote, it seems from my perspective your implementation would work roughly just as well with the `logger` package having an explicit / global `SetLogger` method, but I still would refrain from abstracting loggers unless truly needed)

This might be the case if `context.WithValue` didn't create a hierarchy of contexts. But because it does, the creation of [this context](https://github.com/sourcegraph/sourcegraph/pull/2135/files#diff-997d3db0bf522e0f61f382fd3ab72881R96) has no effect on the context that was [passed into it](https://github.com/sourcegraph/sourcegraph/pull/2135/files#diff-997d3db0bf522e0f61f382fd3ab72881R50).

If you used a global variable to capture the logging context:

- Multiple concurrent requests / function calls would constantly be clobbering one-another's contexts
- Even if you were single threaded/go-routined as this is, logging context from function calls would unnecessarily (and inaccurately) pollute the overall logging context that is used. 

The only way I can see to solve the problem I'm trying to address here would be to pass in a `logCtx` alongside `ctx` to every function call. And, given that the implementation of the logger package is agnostic to whether it's been initialized in any way, this just seems really inelegant.

File: cmd/creaper/build.sh [hunk 7]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
```
[the remark this answers is not in the document]
Comment (goddardcm) [reply cmd/creaper/build.sh]: Makes sense. Let me grind on this a bit and see what I can do.

File: cmd/creaper/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,4 @@
+FROM alpine:3.8
```
[the remark this answers is not in the document]
Comment (goddardcm) [reply cmd/creaper/Dockerfile]: Roger. Will remove this when we get to a solution for your comment on build.sh.

File: cmd/cachereaper/main.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	if *recoveryFactor < 0 || *recoveryFactor > 100 {
+		flag.Usage()
+		os.Exit(2)
+	}
+
+	// Unless force is specified, do some sanity checking.
+	if !*force {
+		if len(*cacheDir) == 0 {
```
Comment (nicksnyder) [inline cmd/cachereaper/main.go]: It seems like this check should happen regardless of `force`?

File: cmd/cachereaper/main.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+			flag.Usage()
+			os.Exit(2)
+		}
+
+		if !strings.HasPrefix(*cacheDir, os.TempDir()) {
```
Comment (nicksnyder) [inline cmd/cachereaper/main.go]: What is the motivation for this check? To prevent a developer from nuking their own machine?

File: cmd/creaper/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicksnyder) [reply cmd/creaper/main.go]: My general opinion is to not add complexity that doesn't have an immediate benefit. If another developer needs to add complexity down the line (e.g. graceful shutdown), I think it would be easier for them to start from scratch and build to their requirements than to look at older code and wonder why this code exists or how it was intended to work (unless there was already an example of graceful shutdown logic to follow).

For this case specifically, it is hard for me to image a future scenario where this service would need to gracefully shutdown (since it only deletes things), so I would be in favor of removing it.

File: cmd/creaper/creaper/creaper.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (slimsag) [reply cmd/creaper/creaper/creaper.go]: Yeah, not worth solving right now probably. Was mostly just curious if/how you thought of this :)

File: cmd/creaper/main.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (slimsag) [reply cmd/creaper/main.go]: I agree with what Nick said above, but I am also OK leaving it as-is. I can see an argument for both cases. My main argument against removing it at this point would be that it does not match our other services consistency-wise.

File: cmd/creaper/creaper/creaper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (slimsag) [reply cmd/creaper/creaper/creaper.go]: Yeah, I checked out the library as well before, it definitely makes sense to use it given we want file access time here.

File: cmd/creaper/logger/logger.go (hunk 1 above)
Comment (slimsag) [reply cmd/creaper/logger/logger.go]: When I had written my comment, I missed the fact that you were in fact passing contextual data here in the form of additional logging data. I can see why you came to these conclusions and it is reasonable to me that you got this implementation.

I would argue that we do not need that extra contextual data in logging, because after so many nested calls in larger apps it quickly becomes tedious to read log lines.

You might really like to check out opentracing, which we use throughout the rest of our app. It has to be hooked up to something external like lightstep or jaeger, but it works really nicely and allows you to include more contextual information for debugging purposes.

As far as keeping this in or not, I am fine merging this either way. Long term, I think it would make sense to use opentracing for the debuggability aspect of this and stick to regular log15 calls with less context about the request in them.

**Comment by slimsag** [comment]
I haven't had a chance to review the recent things here, or even respond to all of this yet -- but I responded to some specific points and I'll try to follow up more next week. Sorry for the delay! :)

**Comment by slimsag** [comment]
I think our CLA bot is dying :/

**Comment by goddardcm** [comment]
I think it just misses me. ;)

Will try to address outstanding issues this weekend. Things have been
hectic.

—c.
